state/address.go

```diff
@@ -88,57 +89,161 @@ func (st *State) Addresses() ([]string, error) {
+	err := mc.Find(bson.D{{"_id", key}}).One(&extantHostPortDoc)
+	if err != nil {
+		if err == mgo.ErrNotFound {
+			return []txn.Op{{
+				C:      controllersC,
+				Id:     key,
+				Insert: bson.D{{"apihostports", fromNetworkHostsPorts(newHostPorts)}},
+			}}, nil
+		}
+		return ops, err
+	}
+
+	// Queue an update operation if the host/port collections differ.
+	extantHostPorts := networkHostsPorts(extantHostPortDoc.APIHostPorts)
+	if !hostsPortsEqual(newHostPorts, extantHostPorts) {
+		ops = []txn.Op{{
+			C:  controllersC,
+			Id: key,
+			Assert: bson.D{{
+				"txn-revno", extantHostPortDoc.TxnRevno,
+			}},
+			Update: bson.D{{
+				"$set", bson.D{{"apihostports", fromNetworkHostsPorts(newHostPorts)}},
+			}},
+		}}
+		logger.Debugf("setting %s: %v", key, newHostPorts)
+	}
+	return ops, nil
+}
+
+// filterHostPortsForManagementSpace filters the collection of API addresses
+// based on the configured management space for the controller.
+// If there is no space configured, or if the one of the slices is filtered down
+// to zero elements, just use the unfiltered slice for safety - we do not
+// want to cut off communication to the controller based on erroneous config.
+func (st *State) filterHostPortsForManagementSpace(apiHostPorts [][]network.HostPort) ([][]network.HostPort, error) {
+	config, err := st.ControllerConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	var hostPortsForAgents [][]network.HostPort
+	if mgmtSpace := config.JujuManagementSpace(); mgmtSpace != "" {
+		hostPortsForAgents = make([][]network.HostPort, len(apiHostPorts))
+		sp := network.SpaceName(mgmtSpace)
+		for i := range apiHostPorts {
+			if filtered, ok := network.SelectHostsPortBySpaces(apiHostPorts[i], sp); ok {
+				hostPortsForAgents[i] = filtered
+			} else {
+				hostPortsForAgents[i] = apiHostPorts[i]
+			}
+		}
+	} else {
+		hostPortsForAgents = apiHostPorts
+	}
+	return hostPortsForAgents, nil
+}
+
+// APIHostPorts returns the collection of *all* API addresses as set by SetAPIHostPorts.
 func (st *State) APIHostPorts() ([][]network.HostPort, error) {
+	hp, err := st.apiHostPortsForKey(apiHostPortsKey)
+	if err != nil {
+		err = errors.Trace(err)
+	}
+	return hp, err
+}
+
+// APIHostPortsForAgents returns the collection of API addresses that should be used
+// by agents.
+// If there is no management network space configured for the controller
+// or if the space is misconfigured, the return will be the same as APIHostPorts.
+// Otherwise the returned addresses will correspond with the management net space.
+// If there is no document at all, we simply fall back to APIHostPorts.
+func (st *State) APIHostPortsForAgents() ([][]network.HostPort, error) {
+	hp, err := st.apiHostPortsForKey(apiHostPortsForAgentsKey)
+	if err != nil {
+		if err == mgo.ErrNotFound {
+			logger.Debugf("No document for %s; using %s", apiHostPortsForAgentsKey, apiHostPortsKey)
+			return st.APIHostPorts()
+		}
+		return nil, errors.Trace(err)
+	}
+	return hp, nil
+}
```

Comment:
I know we have a habit of shoving everything onto State, but we all are kind of unhappy about it, and it makes it hard to test.
Thoughts on breaking it out into a small helper struct that just deals with APIHostPorts and APIHostPortsForAgents and uses interfaces to get the existing docs and comparing them ?
Probably overkill for this patch, but it might help push us towards a more manageable core.